src/circuit/v2/hop.ts

```diff
@@ -0,0 +1,221 @@
+  const { pbstr, request } = options
+  log('received hop message')
+  switch (request.type) {
+    case HopMessage.Type.RESERVE: await handleReserve(options); break
+    case HopMessage.Type.CONNECT: await handleConnect(options); break
+    default: {
+      log.error('invalid hop request type %s via peer %s', options.request.type, options.connection.remotePeer)
+      pbstr.pb(HopMessage).write({ type: HopMessage.Type.STATUS, status: Status.UNEXPECTED_MESSAGE })
+    }
+  }
+}
+
+export async function reserve (connection: Connection): Promise<Reservation> {
+  log('requesting reservation from %s', connection.remotePeer)
+  const stream = await connection.newStream([RELAY_V2_HOP_CODEC])
+  const pbstr = pbStream(stream)
+  const hopstr = pbstr.pb(HopMessage)
+  hopstr.write({ type: HopMessage.Type.RESERVE })
+
+  let response: HopMessage
+  try {
+    response = await hopstr.read()
+  } catch (e: any) {
+    log.error('error passing reserve message response from %s because', connection.remotePeer, e.message)
+    stream.close()
+    throw e
+  }
+
+  if (response.status === Status.OK && (response.reservation != null)) {
+    return response.reservation
+  }
+  const errMsg = `reservation failed with status ${response.status ?? 'undefined'}`
+  log.error(errMsg)
+  throw new Error(errMsg)
+}
+
+const isRelayAddr = (ma: Multiaddr): boolean => ma.protoCodes().includes(CIRCUIT_PROTO_CODE)
+
+async function handleReserve ({ connection, pbstr, relayPeer, relayAddrs, limit, acl, reservationStore }: HopProtocolOptions): Promise<void> {
+  const hopstr = pbstr.pb(HopMessage)
+  log('hop reserve request from %s', connection.remotePeer)
+
+  if (isRelayAddr(connection.remoteAddr)) {
+    log.error('relay reservation over circuit connection denied for peer: %p', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+    return
+  }
+
+  if ((await acl?.allowReserve?.(connection.remotePeer, connection.remoteAddr)) === false) {
+    log.error('acl denied reservation to %s', connection.remotePeer)
+    hopstr.write({ type: HopMessage.Type.STATUS, status: Status.PERMISSION_DENIED })
+    return
+  }
+
+  const result = await reservationStore.reserve(connection.remotePeer, connection.remoteAddr)
+
+  if (result.status !== Status.OK) {
+    hopstr.write({ type: HopMessage.Type.STATUS, status: result.status })
+    return
+  }
+
+  try {
+    hopstr.write({
+      type: HopMessage.Type.STATUS,
+      status: Status.OK,
+      reservation: await makeReservation(relayAddrs, relayPeer, connection.remotePeer, BigInt(result.expire ?? 0)),
+      limit
+    })
+    log('sent confirmation response to %s', connection.remotePeer)
+  } catch (err) {
+    log.error('failed to send confirmation response to %s', connection.remotePeer)
+    await reservationStore.removeReservation(connection.remotePeer)
+  }
+
+  // TODO: how to ensure connection manager doesn't close reserved relay conn
```

Comment:
> how to ensure connection manager doesn't close reserved relay conn

This can be done by tagging the relay peer with a value > 0

Reply:
Are you refering to https://github.com/libp2p/js-libp2p/blob/master/src/connection-manager/index.ts#L603 ?

Reply:
I'm referring to [peerStore.tagPeer](https://libp2p.github.io/js-libp2p-interfaces/interfaces/_libp2p_interface_peer_store.PeerStore.html#tagPeer), I think the `TODO:` is referring to that method in the connection manager.

Reply:
Implementing now